File: libakaza/src/tinylisp.cc
#include <memory>
#include <codecvt>
#include <locale>

#include "../include/tinylisp.h"

using namespace akaza::tinylisp;

static std::shared_ptr<Node> builtin_string_concat(std::vector<std::shared_ptr<Node>> &expr) {
    auto a = expr[0];
    auto b = expr[1];
    std::wstring a_str = dynamic_cast<StringNode *>(&*a)->str();
    std::wstring b_str = dynamic_cast<StringNode *>(&*b)->str();
    return std::shared_ptr<Node>(new StringNode(a_str + b_str));
}

static std::shared_ptr<Node> builtin_current_datetime(std::vector<std::shared_ptr<Node>> &expr) {
    time_t rawtime;
    time(&rawtime);
    struct tm *timeinfo = localtime(&rawtime);
    return std::shared_ptr<Node>(new PointerNode(timeinfo));
}

static std::shared_ptr<Node> builtin_strftime(std::vector<std::shared_ptr<Node>> &expr) {
    std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> cnv;
    auto dt = expr[0];
    auto fmt = expr[1];
    std::wstring fmt_str = dynamic_cast<StringNode *>(&*fmt)->str();
    size_t len = fmt_str.size() * 4;
    char *buffer = new char[len];
    size_t got_len = std::strftime(buffer, len, cnv.to_bytes(fmt_str).c_str(),
                                   static_cast<const struct tm *>(dynamic_cast<PointerNode *>(&*dt)->ptr()));
    auto retval = std::shared_ptr<Node>(new StringNode(cnv.from_bytes(std::string(buffer, got_len))));
    delete[] buffer;
    return retval;
}


std::shared_ptr<Node> TinyLisp::eval(std::shared_ptr<Node> x) const {
    if (x->type() == NODE_SYMBOL) {
        std::wstring symbol = dynamic_cast<SymbolNode *>(&*x)->symbol();
        if (symbol == L"current-datetime") {
            return std::shared_ptr<Node>(new FunctionNode(builtin_current_datetime));
        } else if (symbol == L"strftime") {
            return std::shared_ptr<Node>(new FunctionNode(builtin_strftime));
        } else if (symbol == L".") {
            return std::shared_ptr<Node>(new FunctionNode(builtin_string_concat));
        } else {
            std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> cnv;
            throw std::runtime_error(cnv.to_bytes(std::wstring(L"Unknown function: ") + symbol));
        }
    } else if (x->type() == NODE_LIST) { // (proc exp*)
        std::vector<std::shared_ptr<Node>> exps;
        for (auto &exp : *dynamic_cast<ListNode *>(&*x)->children()) {
            exps.push_back(eval(exp));
        }

        auto proc = exps[0];
        exps.erase(exps.begin());
        return dynamic_cast<FunctionNode *>(&*proc)->call(exps);
    } else {
        return x;
    }
}

std::shared_ptr<Node>
TinyLisp::_read_from(std::vector<std::wstring> &tokens, int depth, const std::wstring &src) const {
    if (tokens.empty()) {
        std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> cnv;
        throw std::runtime_error(std::string("Unexpected EOF while reading(LISP): ") + cnv.to_bytes(src));
    }
    std::wstring token = tokens[0];
    tokens.erase(tokens.begin());
    if (token == L"(") {
        std::vector<std::shared_ptr<Node>> values;
        while (tokens[0] != L")") {
            values.push_back(_read_from(tokens, depth + 1, src));
        }
        tokens.erase(tokens.begin()); // pop off ")"
        return std::shared_ptr<Node>(new ListNode(values));
    } else if (token == L")") {
        throw std::runtime_error("Unexpected ')'");
    } else {
        return _atom(token);
    }
}

std::shared_ptr<Node> TinyLisp::_atom(const std::wstring &token) {
    if (!token.empty() && token[0] == '"') {
        return std::shared_ptr<Node>(
                new StringNode(token.substr(1, token.size() - 2)));
    } else {
        return std::shared_ptr<Node>(new SymbolNode(token));
    }
}

File: akaza-core/libakaza/src/tinylisp.rs
use std::borrow::Borrow;
use std::collections::VecDeque;

use anyhow::{anyhow, Result};
use chrono::{DateTime, Local};

/**
 * 簡易 LISP のインタープリタ for akaza。
 * ref. http://norvig.com/lispy.html
 */

type FunctionCallback = fn(args: VecDeque<TinyLispNode>) -> Result<TinyLispNode>;

#[derive(Clone, PartialEq, Debug)]
enum TinyLispNode {
    List(Vec<TinyLispNode>),
    String(String),
    Symbol(String),
    Function(FunctionCallback),
    LocalDateTime(DateTime<Local>),
}

fn dump_node(node: &TinyLispNode, depth: i32) {
    for _i in 0..depth {
        print!(" ");
    }

    match node {
        TinyLispNode::List(list) => {
            println!("ListNode:");
            for item in list {
                dump_node(item, depth + 1);
            }
        }
        TinyLispNode::String(s) => {
            println!("StringNode({})", s);
        }
        TinyLispNode::Symbol(s) => {
            println!("SymbolNode({})", s);
        }
        TinyLispNode::Function(_) => {
            println!("FunctionNode()");
        }
        TinyLispNode::LocalDateTime(_) => {
            println!("LocalDateTime()");
        }
    }
}

fn builtin_string_concat(args: VecDeque<TinyLispNode>) -> Result<TinyLispNode> {
    let a = &args[0];
    let b = &args[1];

    let TinyLispNode::String(a_str) = a else {
        return Err(anyhow!("argument for '.' operator should be string."));
    };
    let TinyLispNode::String(b_str) = b else {
        return Err(anyhow!("argument for '.' operator should be string."));
    };

    Ok(TinyLispNode::String(a_str.clone() + b_str))
}

fn builtin_current_datetime(_args: VecDeque<TinyLispNode>) -> Result<TinyLispNode> {
    Ok(TinyLispNode::LocalDateTime(Local::now()))
}

fn builtin_strftime(args: VecDeque<TinyLispNode>) -> Result<TinyLispNode> {
    let dt = &args[0];
    let fmt = &args[1];
    let TinyLispNode::LocalDateTime(dt) = dt else {
        return Err(anyhow!("1st argument of strftime should be LocalDateTime"));
    };
    let TinyLispNode::String(fmt) = fmt else {
        return Err(anyhow!("2nd argument of strftime should be string"));
    };
    let got = dt.format(fmt).to_string();
    Ok(TinyLispNode::String(got))
}

struct TinyLisp {}

impl TinyLisp {
    pub fn run(sexp: &str) -> Result<String> {
        let parsed = Self::parse(sexp);
        let parsed = match parsed {
            Ok(node) => node,
            Err(err) => return Err(err),
        };
        let result = Self::eval(&parsed);
        return match result {
            Ok(node) => {
                let node = node.borrow();
                if let TinyLispNode::String(ret) = node {
                    Ok(ret.clone())
                } else {
                    Err(anyhow!("Result of LISP must be String"))
                }
            }
            Err(err) => Err(err),
        };
    }

    fn parse(sexp: &str) -> Result<TinyLispNode> {
        let mut tokens = Self::tokenize(sexp);

        Self::_read_from(&mut tokens, 0)
    }

    fn eval(node: &TinyLispNode) -> Result<TinyLispNode> {
        match node {
            TinyLispNode::Symbol(symbol) => {
                if symbol == "." {
                    Ok(TinyLispNode::Function(builtin_string_concat))
                } else if symbol == "current-datetime" {
                    Ok(TinyLispNode::Function(builtin_current_datetime))
                } else if symbol == "strftime" {
                    Ok(TinyLispNode::Function(builtin_strftime))
                } else {
                    Err(anyhow!("Unknown function: {}", symbol))
                }
            }
            TinyLispNode::List(list) => {
                let mut exps: VecDeque<TinyLispNode> = VecDeque::new();
                for exp in list {
                    let result = Self::eval(exp);
                    match result {
                        Ok(node) => {
                            exps.push_back(node);
                        }
                        Err(err) => {
                            return Err(err);
                        }
                    }
                }
                let Some(proc) = exps.pop_front() else {
                    return Err(anyhow!("Empty list."));
                };
                if let TinyLispNode::Function(proc) = proc {
                    proc(exps)
                } else {
                    Err(anyhow!("Expected function... But it's not."))
                }
            }
            _ => Ok(node.clone()),
        }
    }

    fn tokenize(buf: &str) -> VecDeque<String> {
        // TODO This method should care the string literal that contains space character.
        let buf = buf.replace('(', " ( ");
        let buf = buf.replace(')', " ) ");
        let tokens: Vec<&str> = buf.split(' ').collect();
        return tokens
            .iter()
            .filter(|t| !t.is_empty())
            .map(|f| f.to_string())
            .collect();
    }

    fn _read_from(tokens: &mut VecDeque<String>, _depth: i32) -> Result<TinyLispNode> {
        if tokens.is_empty() {
            return Err(anyhow!("Unexpected EOF while reading(LISP)"));
        }

        let Some(token) = tokens.pop_front() else {
            return Err(anyhow!("Missing token... Unexpected EOS."));
        };
        if token == "(" {
            let mut values: Vec<TinyLispNode> = Vec::new();
            while tokens[0] != ")" {
                let result = Self::_read_from(tokens, _depth + 1);
                match result {
                    Ok(node) => values.push(node),
                    Err(err) => return Err(err),
                }
            }
            tokens.pop_front();
            Ok(TinyLispNode::List(values))
        } else if token == ")" {
            Err(anyhow!("Unexpected token: ')'"))
        } else {
            Ok(Self::_atom(&token))
        }
    }

    fn _atom(token: &String) -> TinyLispNode {
        return if !token.is_empty() && token.starts_with('\"') {
            TinyLispNode::String(
                token
                    .strip_prefix('\"')
                    .unwrap()
                    .strip_suffix('\"')
                    .unwrap()
                    .to_string(),
            )
        } else {
            TinyLispNode::Symbol(token.clone())
        };
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_atom() {
        // symbol node
        {
            let node = TinyLisp::_atom(&"hogehoge".to_string());
            assert_eq!(node, TinyLispNode::Symbol("hogehoge".to_string()));
        }

        // string node
        {
            let node = TinyLisp::_atom(&"\"hogehoge\"".to_string());
            assert_eq!(node, TinyLispNode::String("hogehoge".to_string()));
        }
    }

    #[test]
    fn test_tokenize() {
        let tokens = TinyLisp::tokenize("(. \"a\" \"b\")");
        assert_eq!(tokens.len(), 5);
        assert_eq!(tokens[0], "(".to_string());
        assert_eq!(tokens[1], ".".to_string());
        assert_eq!(tokens[2], "\"a\"".to_string());
        assert_eq!(tokens[3], "\"b\"".to_string());
        assert_eq!(tokens[4], ")".to_string());
    }

    #[test]
    fn test_run() {
        let p = TinyLisp::run("\"hoge\"").unwrap();
        assert_eq!(p, "hoge");
    }

    #[test]
    fn test_builtin_concat() {
        let p = TinyLisp::run("(. \"h\" \"b\")").unwrap();
        assert_eq!(p, "hb");
    }

    #[test]
    fn test_builtin_strftime() {
        let src = "(strftime (current-datetime) \"%Y-%m-%d\")".to_string();
        let parsed = TinyLisp::parse(&src).unwrap();
        dump_node(&parsed, 0);
        let p = TinyLisp::run(&src).unwrap();
        assert!(p.starts_with('2')); // this test succeeds until year of 2999.
    }
}
